src/stan/variational/advi.hpp

```diff
@@ -150,18 +150,143 @@ namespace stan {
+            if (iter_tune == 1) {
+              params_prop += elbo_grad.square();
+            } else {
+              params_prop = pre_factor * params_prop +
+                            post_factor * elbo_grad.square();
+            }
+            eta_scaled = eta / sqrt(static_cast<double>(iter_tune));
+
+            // Stochastic gradient update
+            variational += eta_scaled * elbo_grad / (tau + params_prop.sqrt());
+          }
+          elbo = calc_ELBO(variational);
+
+          // Check if:
+          // (1) ELBO at current eta is worse than the best ELBO
+          // (2) the best ELBO hasn't actually diverged
+          if (elbo < elbo_best && elbo_best > elbo_init) {
+            if (print_stream_) {
+              *print_stream_ << "Success!"
+                << " Found best tuned hyperparameter"
+                << " [eta = " << eta_best
+                << "]";
+              if (eta_sequence.size() > 0) {
+                *print_stream_
+                  << " earlier than expected.";
+              }
+              *print_stream_ << std::endl << std::endl;
+            }
+            do_more_tuning = false;
+          } else {
+            if (eta_sequence.size() > 0) {
+              // Restart phase
+              elbo_best = elbo;
+              eta_best = eta;
+            } else {
+              // No more eta values to try, so use current eta if it
+              // didn't diverge or fail if it did diverge
+              if (elbo > elbo_init) {
+                if (print_stream_)
+                  *print_stream_ << "Success!"
+                    << " Found best tuned hyperparameter"
+                    << " [eta = " << eta_best
+                    << "]" << std::endl << std::endl;
+                eta_best = eta;
+                do_more_tuning = false;
+              } else {
+                const char* name = "All proposed step-sizes";
+                const char* msg1 = "failed. Your model may be either "
+                  "severely ill-conditioned or misspecified.";
+                stan::math::domain_error(function, name, "", msg1);
+                return 0;
+              }
+            }
+            // Reset
+            params_prop = Q(model_.num_params_r());
+          }
+          variational = Q(cont_params_);
+        }
+        return eta_best;
+      }
+
+      /**
+       * Runs stochastic gradient ascent with an adaptive stepsize.
        *
        * @param  variational    variational distribution
+       * @param  eta            eta parameter for stepsize scaling
        * @param  tol_rel_obj    relative tolerance parameter for convergence
        * @param  max_iterations max number of iterations to run algorithm
        */
-      void robbins_monro_adagrad(Q& variational,
-                                 double tol_rel_obj,
-                                 int max_iterations) const {
+      void stochastic_gradient_ascent(Q& variational,
```

Comment:
I'd like to see it pulled out as its own standalone tested function
that doesn't depend on the variational algorithm.  And why isn't Q a const
here?  Are you resetting it in the algorithm?


> On Aug 26, 2015, at 9:04 AM, Daniel Lee <notifications@github.com> wrote:
> 
> In src/stan/variational/advi.hpp:
> 
> >         * @param  tol_rel_obj    relative tolerance parameter for convergence
> >         * @param  max_iterations max number of iterations to run algorithm
> >         */
> > -      void robbins_monro_adagrad(Q& variational,
> > -                                 double tol_rel_obj,
> > -                                 int max_iterations) const {
> > +      void stochastic_gradient_ascent(Q& variational,
> 
> add unit test for stochastic_gradient_ascent
> 
> —
> Reply to this email directly or view it on GitHub.
>